pkg/node/node.go

```diff
@@ -424,41 +429,109 @@ func (n *Node) publishSignature(
-	conflictingPublicKey := make(chan *eth.ConflictingPublicKeySubmittedEvent)
+	currentBlock, err := n.ethereumChain.BlockCounter().CurrentBlock()
+	if err != nil {
+		logger.Errorf(
+			"could not get current block while confirming "+
+				"signature submission for keep [%s]: [%v]",
+			keepAddress.String(),
+			err,
+		)
+		return false
+	}
 
-	subscriptionPublicKeyPublished, err := n.ethereumChain.OnPublicKeyPublished(
-		keepAddress,
-		func(event *eth.PublicKeyPublishedEvent) {
-			publicKeyPublished <- event
+	isSignatureConfirmed, err := chainutil.WaitForBlockConfirmations(
+		n.ethereumChain.BlockCounter(),
+		currentBlock,
+		blockConfirmations,
+		func() (bool, error) {
+			isAwaitingSignature, err := n.ethereumChain.IsAwaitingSignature(
+				keepAddress,
+				digest,
+			)
+			if err != nil {
+				return false, err
+			}
+
+			return !isAwaitingSignature, nil
 		},
 	)
 	if err != nil {
 		logger.Errorf(
-			"failed on watching public key published event for keep [%s]: [%v]",
+			"could not confirm signature submission for keep [%s]: [%v]",
 			keepAddress.String(),
 			err,
 		)
+		return false
+	}
+
+	if !isSignatureConfirmed {
+		logger.Errorf(
+			"signature submission for keep [%s] not confirmed; "+
+				"trying to submit the signature again",
+			keepAddress.String(),
+		)
+		return false
 	}
 
+	logger.Infof(
+		"signature submission for keep [%s] has been confirmed",
+		keepAddress.String(),
+	)
+
+	return true
+}
+
+// monitorKeepPublicKeySubmission observes the chain until either the first
+// conflicting public key is published or until keep established public key
+// or until the key generation timed out. It also tries to re-submit the public
+// key if it has not been submitted in the given time frame as a way to protect
+// against chain forks (i.e. transaction mined in a fork that has been dropped).
+//
+// When event informing about conflicting public key is emitted from the chain,
+// monitoring stops immediately. Even if this event was emitted in a minority
+// fork, it is clear that the operator who submitted the conflicting key is
+// dishonest and it is better to abandon this keep.
+//
+// Function checks the chain periodically and inspects whether the public key
+// has been registered on the chain. If the key is there, the function waits
+// for the required number of confirmations and repeats the check. If the public
+// key for the keep is still there, the monitoring exits successfully.
+// In the case when public key for the keep is not yet established during the
+// periodic check or it is gone after waiting for a certain number of
+// confirmations (chain reorganization), this function will attempt to submit
+// the public key again.
+func (n *Node) monitorKeepPublicKeySubmission(
+	keepAddress common.Address,
+	publicKey [64]byte,
+) {
+	monitoringCtx, monitoringCancel := context.WithTimeout(
+		context.Background(),
+		monitorKeepPublicKeySubmissionTimeout,
+	)
+	defer monitoringCancel()
```

Comment:
I am afraid there is one thing we did not take into account: retried `SubmitPublicKey` transactions are not failing on gas estimate but they are getting mined and are burning gas on a failed transaction. For `SubmitPublicKey`, the gas estimate is hardcoded in `ethereum` implementation. If it is good or not, that's another discussion but it becomes very clear now than in this code we should not make assumptions about how the implementation works.

The scenario I was testing:
Three nodes in the group and one node exited just before submitting the public key but this can be also a fatal Ethereum client failure or any other issue with it. Two other nodes kept trying to submit their public keys, burning `29447` gas on each attempt:
```
2020-11-18T16:58:47.712+0100	INFO	keep-ecdsa	keep [0x8914a0983C0C9803Bcf70B13C132A8f9B41F3D4C] still does not have a confirmed public key; re-submitting public key [c4d5a7e8702776e2da4ac1c9873f9c07ef8c99af1c22da27e2c7b0ae868c78384fd7b6b947a627aaa40950372692e4850d1f021457809c4bcacb528d081dd335]
2020-11-18T16:58:48.320+0100	INFO	keep-contract-BondedECDSAKeep	submitted transaction submitPublicKey with id: [0x2c4e0881f5bd9561fa3fa0332e0c4cc6c3368c294315b5ac6a3f09b907d7ac66] and nonce [170]
2020-11-18T16:58:51.325+0100	INFO	keep-ethutil	transaction [2c4e08…d7ac66] mined with status [0] at block [1220]
```

Depending on the gas price and block times, those clients can burn more ether on those failed transactions than it would be seized from them for not publishing a keep public key.

I was considering whether we can:
- remove `monitoringCtx` altogether,
- change `pubkeyCheckTrigger` to tick on time, say every 20 minutes,
- give up after two executions of `case pubkeyCheckBlock := <-pubkeyCheckTrigger`.

What I like about this solution is that we have a fixed number of retries so there is a limit of potentially failed transactions and we use time as a ticker, just like in the monitoring of signature submission, so we are consistent.

WDYT?

Reply:
Yep, sounds good. I'll change that in the way you proposed.

Reply:
I changed but I introduced a constant 10 minutes trigger with a give up after six executions. This is simpler than juggling with two different tickers or weird counters. Also, this change opened the possibility to add `continue` in places you mentioned in this comment https://github.com/keep-network/keep-ecdsa/pull/610#discussion_r526021630

Reply:
The reason I proposed 5/20/20 checks is to have the first check executed pretty soon so that in the optimistic case - and I assume this is the case for 95% of keeps - the log about the public key confirmed is closer to logs about generating this key. 

10 minutes is also fine but I would consider reducing the number of retries to 3. As you previously mentioned, `SubmitPublicKey` has retries implemented under the hood and here we are just protecting against reorgs. In the case of one client in the group not operating properly, reducing the number of retries from 6 to 3 should let two other clients save gas. And 3 retries protecting against reorgs should be more than enough given that we are currently not retrying and in the vast majority of cases we are fine.